### DISCUSSION TITLE
Async Channel in Device Service

### DISCUSSION THREAD

**Discussion by sounak-1989** [edgexfoundry/community#31] [General]
Hi,

I needed some help in Device Service SDK. I wanted to get a Resource defined in the profile but I want to use another resource as parameter for retrieving the main resource. For example I have an IO device which has multiple temperature sensors attached to it so I have defined 2 resources namely
- PortId
- Temperature
I want to read the Temperature but pass the PortId as a parameter to specify which port I want to read. So far from what I understand from the device service example I have 2 options
- Design my own REST API system which takes care of it as shown in the device-rest-go
- Use a Write Command where I pass PortId as the first parameter and the CommandName as the second parameter. 
I don't want to design an API router from scratch so I went the second way based on the example from device-llrp-go. 
But the problem I am facing is even if I save the result in the async Channel in the driver struct I am just getting a 200OK in the response instead of the result that I want to return.

Any idea how I can access the result from the async channel? Thanks

**Comment by sounak-1989** [comment]
Hey guys,
Found the answer to my own question after some research. The Async Channel values are sent as events to core-data. When I requested the event log for that device I could see the values I requested in my command.

**Comment by cloudxxx8** [comment]
When you implement the Device Services based on Device Service SDK, you can retrieve the resource information from `attributes` in the profile.
They will be in `CommandRequest`
https://github.com/edgexfoundry/device-sdk-go/blob/7d510a9be1a3cd2158385898861d08f1c1b53357/pkg/models/protocoldriver.go#L30-L38
https://github.com/edgexfoundry/device-sdk-go/blob/7d510a9be1a3cd2158385898861d08f1c1b53357/pkg/models/commandrequest.go#L15

In the async value handler, you can still use the internal DS API to retrieve the Device or Profile from cache.
https://github.com/edgexfoundry/device-sdk-go/blob/7d510a9be1a3cd2158385898861d08f1c1b53357/pkg/interfaces/service.go#L62
For example:
https://github.com/edgexfoundry/device-virtual-go/blob/0c2f1e9213ddf10d1b3316c025c729eab3a2bc23/internal/driver/virtualdriver.go#L161

**Reply by sounak-1989** [reply]
@cloudxxx8 This is great. I was actually looking for some examples like this. Thanks a lot for directing me in that direction.
